compress_multiphoton/compression.py

```diff
@@ -2,50 +2,84 @@
-
-    # Select pixels in good range
-    pixels = pixels[pixels_mask]
-    unique_pixels, counts = np.unique(pixels, return_counts=True)
-
-    # Compute noise variance
-    variances = ((scan[:, :, :-1] - scan[:, :, 1:]) ** 2 / 2)[pixels_mask]
-    pixels -= min_intensity
-    variance_sum = np.zeros(len(unique_pixels))  # sum of variances per pixel value
-    for i in range(0, len(pixels), int(1e8)):  # chunk it for memory efficiency
-        variance_sum += np.bincount(
-            pixels[i : i + int(1e8)], weights=variances[i : i + int(1e8)], minlength=np.ptp(unique_pixels) + 1
-        )[unique_pixels - min_intensity]
-    unique_variances = variance_sum / counts  # average variance per intensity
-
-    # Compute quantal size (by fitting a linear regressor to predict the variance from intensity)
-    X = unique_pixels.reshape(-1, 1)
-    y = unique_variances
-    model = TheilSenRegressor()  # robust regression
-    model.fit(X, y)
+def _longest_run(bool_array):
+    step = np.diff(np.int8(bool_array), prepend=0, append=0)
+    on = np.where(step == 1)[0]
+    off = np.where(step == -1)[0]
+    i = np.argmax(off - on)
+    return slice(on[i], off[i])
+
+
+def compute_quantal_size(movie: np.array) -> dict:
+    """Calculate quantal size for a movie.
+
+    Args:
+        movie (np.array):  A movie in the format (height, width, time).
+
+    Returns:
+        output (dict): A dictionary with the following keys:
+            - 'model': The fitted TheilSenRegressor model.
+            - 'min_intensity': Minimum intensity used.
+            - 'max_intensity': Maximum intensity used.
+            - 'intensity_levels': Intensity levels.
+            - 'variance': Variances at intensity levels.
+            - 'quantal_size': Estimated quantal size.
+            - 'zero_level': Estimated ADC gain.
+    """
+    movie = movie.astype(np.int32, copy=False)
+    intensity = (movie[:, :, :-1] + movie[:, :, 1:] + 1) // 2
+    difference = movie[:, :, :-1] - movie[:, :, 1:]
+
+    CTS_THRESHOLD = 1000
+    cts = np.bincount(intensity.flatten())
+    cts_slice = _longest_run(cts > CTS_THRESHOLD)
+    cts_slice = slice(max(cts_slice.stop * 20 // 100, cts_slice.start), cts_slice.stop)
+
+    counts = counts[cts_slice]
+    idx = (intensity >= cts_slice.start) & (intensity < cts_slice.stop)
+    variance = (
+        np.bincount(
+            intensity[idx] - cts_slice.start,
+            weights=(np.float32(difference[idx]) ** 2) / 2,
+        )
+        / counts
+    )
+
+    intensity_levels = np.arange(cts_slice.start, cts_slice.stop)
+
+    model = TheilSenRegressor()
+    model.fit(intensity_levels.reshape(-1, 1), variance)
     quantal_size = model.coef_[0]
     zero_level = -model.intercept_ / model.coef_[0]
 
-    return (model, min_intensity, max_intensity, unique_pixels, unique_variances, quantal_size, zero_level)
+    output = dict(
+        model=model,
+        min_intensity=cts_slice.start,
+        max_intensity=cts_slice.stop,
```

Comment:
should this be `stop+1` if you literally mean to indicate the max_intensity? 

What's the intention for min and max? Because these specify the range we chose to fit, not the entire range.

Reply:
One reason I can think of is for plotting the best fit curve.

Reply:
I see. Then you can rename since min and max are used differently. Or make `max_intensity=slice.stop-1` to be precise.

Reply:
Done.